File: leetcode/Stack/Python/20. Valid parentheses.py
#https://leetcode.com/submissions/detail/116515590/

#Given a string containing just the characters '(', ')', '{', '}', '[' and ']', 
#determine if the input string is valid.
#The brackets must close in the correct order, "()" and "()[]{}" are all valid 
#but "(]" and "([)]" are not.

class Solution(object):
    
    def isMatch(self, c1, c2):
        if (c1 == '[' and c2 ==']') or (c1 == '(' and c2 == ')') or (c1 == '{' and c2 == '}'):
            return True
        else:
            return False
        
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for c in s:
            if stack and self.isMatch(stack[-1], c):
                stack.pop()
            else:
                stack.append(c)
        
        if stack:
            return False
        else:
            return True

File: leetcode/Stack/Java/20. Valid parentheses.java
//https://leetcode.com/submissions/detail/116353682/

// Given a string containing just the characters '(', ')', '{', '}', '[' and ']', 
// determine if the input string is valid.

// The brackets must close in the correct order, "()" and "()[]{}" are all valid 
// but "(]" and "([)]" are not.



class Solution {
    public boolean isMatch(char c1, char c2){
        if ((c1 == '(' && c2 == ')') || (c1 =='[' && c2 == ']') || (c1 == '{' && c2 == '}')){
            return true;
        }else{
            return false;
        }
    }
    public boolean isValid(String s) {
        Stack stack = new Stack();
        
        //you must close the last before the next
        for(int i= 0; i < s.length() ; i++){
            if(!stack.empty() && isMatch((char)stack.peek(), s.charAt(i))){
                stack.pop();
            }else{
                stack.push(s.charAt(i));    
            }
        }
        
        if (stack.empty())
            return true;
        else
            return false;
        
    }
}